I'm not sure this is supposed to build on darwin.
Here are the logs in case somebody is interested
```
ran test tests/nix-channel.sh... [FAIL]
    + clearProfiles
    + profiles=/private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/var/nix/
    + rm -rf /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/var/nix/pr
    + rm -f /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/test-home/.
    + nix-channel --add http://foo/bar xyzzy
    + nix-channel --list
    + grep -q http://foo/bar
    + nix-channel --remove xyzzy
    warning: selector 'xyzzy' matched no installed derivations
    building '/private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/ic3
    + '[' -e /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/test-home/
    ++ cat /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/test-home/.n
    + '[' '' = '' ']'
    + rm -rf /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/foo
    + mkdir -p /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/foo
    +++ nix-instantiate dependencies.nix
    warning: you did not specify '--add-root'; the result might be removed by the garbage collector
    ++ nix-store -r /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/sto
    these 4 derivations will be built:
      /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/y48v2202l57
      /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/xk7y0123hbb
      /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/xzmp60w4rsg
      /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/2d2kklm7hb6
    building '/private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/xzm
    building '/private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/y48
    building '/private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/xk7
    building '/private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/2d2
    FOO
    warning: you did not specify '--add-root'; the result might be removed by the garbage collector
    + nix copy --to 'file:///private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-cha
    + rm -rf /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/nixexprs
    + mkdir -p /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/nixexprs
    + cp config.nix dependencies.nix dependencies.builder0.sh /private/tmp/nix-build-nix-2.4pre20210513_34
    + ln -s dependencies.nix /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-ch
    + cd /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel
    + tar cvf - nixexprs
    + bzip2
    nixexprs/
    nixexprs/config.nix
    nixexprs/dependencies.builder0.sh
    nixexprs/dependencies.nix
    nixexprs/default.nix
    + nix-channel --add file:///private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-
    + nix-channel --update
    unpacking channels...
    + nix-env -qa '*' --meta --xml --out-path
    + grep -q 'meta.*description.*Random test package' /private/tmp/nix-build-nix-2.4pre20210513_340f831.d
    + grep -q 'item.*attrPath="foo".*name="dependencies-top"' /private/tmp/nix-build-nix-2.4pre20210513_34
    + nix-env -i dependencies-top
    installing 'dependencies-top'
    building '/private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/store/slz
    + '[' -e /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/test-home/
    + clearProfiles
    + profiles=/private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/var/nix/
    + rm -rf /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/var/nix/pr
    + rm -f /private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-channel/test-home/.
    + nix-channel --add file:///private/tmp/nix-build-nix-2.4pre20210513_340f831.drv-1/nix-test/tests/nix-
    + nix-channel --update
    error: value is the built-in function 'import' while a set was expected
    error: program '/nix/store/iz0yqm179zy24rp9pbpdb9ljwi4x4lww-nix-2.4pre20210513_340f831/bin/nix-build'
make: *** [mk/lib.mk:106: tests/nix-channel.sh.test] Error 1
```

--- a/pkgs/tools/package-management/nix/default.nix
+++ b/pkgs/tools/package-management/nix/default.nix
@@ -225,6 +225,13 @@ in rec {
     };
 
     inherit storeDir stateDir confDir boehmgc;
+
+    patches = [
+      (fetchpatch {
+        url = "https://github.com/NixOS/nix/commit/8c7e043de2f673bc355d83f1e873baa93f30be62.patch";
+        sha256 = "sha256-aTcUnZXheewnyCT7yQKnTqQDKS2uDoN9plMQgxJH8Ag=";
+      })
+    ];
   });
 
   nixExperimental = nixUnstable.overrideAttrs (prev: {

